stream/stream_dvbv5.c

```diff
@@ -0,0 +1,577 @@
+        {"timeout", OPT_INT(cfg_timeout), M_RANGE(0, 3600)},
+        {0}
+    },
+    .size = sizeof(struct dvbv5_opts),
+    .defaults = &(const dvbv5_opts_t){
+        .cfg_adapter = 0,
+        .cfg_frontend = 0,
+        .cfg_demux = 0,
+        .cfg_cc = NULL,
+        .n_vpid = 0,
+        .n_apid = 0,
+        .cfg_verbose = 0,
+        .cfg_rec_psi = 0,
+        .cfg_lnb_name = NULL,
+        .cfg_lna = LNA_AUTO,
+        .cfg_sat_number = -1,
+        .cfg_diseqc_wait = 0,
+        .cfg_freq_bpf = 0,
+        .cfg_port = 0,
+        .cfg_server = NULL,
+        .cfg_timeout = 120,
+    },
+};
+static int timeout_flag = 0;
+
+static void *do_timeout(void *priv_ptr) {
+    dvb_priv_t *priv = priv_ptr;
+    usleep(priv->opts->cfg_timeout * 1000000);
+    timeout_flag = 1;
+    return NULL;
+}
+
+static void dvbv5_close(stream_t *stream)
+{
+    dvb_priv_t *priv = stream->priv;
+
+    dvb_dev_free(priv->dvb);
+}
+
+static int dvbv5_stream_control(struct stream *s, int cmd, void *arg)
+{
+    MP_ERR(s, "stream control");
+    switch (cmd) {
+    case STREAM_CTRL_GET_METADATA: {
+        struct mp_tags *metadata = talloc_zero(NULL, struct mp_tags);
+        char *progname = s->path;
+        mp_tags_set_str(metadata, "title", progname);
+        *(struct mp_tags **)arg = metadata;
+        return STREAM_OK;
+    }
+    }
+    return STREAM_UNSUPPORTED;
+}
+
+static int dvbv5_streaming_read(stream_t *stream, void *buffer, int size)
+{
+    int r, pos = 0;
+
+    dvb_priv_t *priv = stream->priv;
+
+    r = dvb_dev_read(priv->dvr_fd, buffer, size);
+
+    pos += r;
+
+    return pos;
+}
+
+/*
+ * This function was adapted for MPV from v4l-utils/utils/dvb/dvbv5-zap.c
+ * version v4l-utils-1.20.0-118-g9a628fbc
+ */
+static int parse(stream_t *stream, struct dvb_v5_fe_parms *parms,
+         int *vpid, int *apid, int *sid)
+{
+    struct mpv_global *global = stream->global;
+    void *talloc_ctx;
+    char *conf_file;
+    char *channel;
+    enum dvb_file_formats input_format;
+    channel = stream->path;
+    dvb_priv_t *priv;
+    struct dvbv5_opts *args;
+
+    priv = stream->priv;
+    args = priv->opts;
+
+    input_format = dvb_parse_format("DVBV5");
+
+    struct dvb_file *dvb_file;
+    struct dvb_entry *entry;
+    int i;
+    uint32_t sys;
+
+    /* This is used only when reading old formats */
+    switch (parms->current_sys) {
+    case SYS_DVBT:
+    case SYS_DVBS:
+    case SYS_DVBC_ANNEX_A:
+    case SYS_ATSC:
+        sys = parms->current_sys;
+        break;
+    case SYS_DVBC_ANNEX_C:
+        sys = SYS_DVBC_ANNEX_A;
+        break;
+    case SYS_DVBC_ANNEX_B:
+        sys = SYS_ATSC;
+        break;
+    case SYS_ISDBT:
+    case SYS_DTMB:
+        sys = SYS_DVBT;
+        break;
+    default:
+        sys = SYS_UNDEFINED;
+        break;
+    }
+    talloc_ctx = talloc_new(NULL);
+    conf_file = mp_find_config_file(talloc_ctx, global, "channels.conf.dvbv5");
+    dvb_file = dvb_read_file_format(conf_file, sys, input_format);
```

Comment:
If the config file does not exist, this fails with `Bad address` in `libdvbv5` (and only then, `-2` is returned, but no clear error message is generated). It might be nicer to check if the file exists first and produce an error message pointing to this error ;-).